### PR TITLE
ENH: Add subspace tracking to ORNMF (consistent with ORPCA)

### PR DESCRIPTION
**Pull request by tjof2**
The ORPCA matrix factorization function supports a SGD method. Using this allows the subspace to be tracked if it changes over time, which is useful for e.g. foreground-background separation in videos.

This change copies the same subspace tracking code over to ONMF, which currently assumes that when `X = WH^T + E`, `W` is stationary for the whole dataset. If your `nsamples` dimension is time, this allows for tracking variations in `W`.

Also, the ONMF algorithm hasn't been accessible from `s.decomposition()` since it was introduced to HyperSpy - this was presumably an oversight.

### CODE REVIEW

**Comment by tjof2** [comment]
This is ready for review

**Comment by francisco-dlp** [comment]
The code looks good, but it'll be good to:

* [ ] Add more information about the feature in the User Guide in the Big Data section.
* [ ] Add a reference to the relevant scientific article
* [ ] Make the ORPCA and ONMF keywords for this feature consistent

It'll be also good to hear from @to266—the author of the ONMF code—about this.

**Comment by tjof2** [comment]
Thanks @francisco-dlp  - I'll work my way through your list!

**Comment by tjof2** [comment]
I'll work on the documentation.

**Comment by tjof2** [comment]
> * [x]  Add more information about the feature in the User Guide in the Big Data section.
> * [X]  Add a reference to the relevant scientific article
> * [x]  Make the ORPCA and ONMF keywords for this feature consistent

**Comment by tjof2** [comment]
This is now ready for review again @francisco-dlp

**Comment by tjof2** [comment]
Or rather, ready for review once I fix the tests...

File: doc/user_guide/mva.rst
Comment (dnjohnstone) [inline doc/user_guide/mva.rst:363]: the Momentum Stochastic Gradient Descent (Momentum SGD) method...

File: doc/user_guide/mva.rst
Comment (dnjohnstone) [inline doc/user_guide/mva.rst]: A more robust alternative is available, although it is typically much slower.

**Review by dnjohnstone — APPROVED** [review]
I think this was basically just waiting for the docs to be checked - I've suggested two very minor changes that could be made easily and then I think this could be merged.

File: doc/user_guide/mva.rst
Comment (tjof2) [reply doc/user_guide/mva.rst:363]: good idea

Comment (tjof2) [reply doc/user_guide/mva.rst:363]: I've defined MomentumSGD higher up now.

**Comment by tjof2** [comment]
> I think this was basically just waiting for the docs to be checked - I've suggested two very minor changes that could be made easily and then I think this could be merged.

Thanks @dnjohnstone for your suggestions!

**Comment by tjof2** [comment]
I don't believe there is anything more to review here?

**Comment by dnjohnstone** [comment]
@francisco-dlp @ericpre - what is this waiting for? Just for conflicts to be resolved?

**Comment by tjof2** [comment]
> @francisco-dlp @ericpre - what is this waiting for? Just for conflicts to be resolved?

I fixed the conflicts @dnjohnstone - hadn't noticed those, thanks - tests still pass

File: hyperspy/learn/ornmf.py
Comment (dnjohnstone) [inline hyperspy/learn/ornmf.py]: Too many "" --> Should be: "Normalization with an iterator is not possible, option ignored."

File: hyperspy/learn/rpca.py
Comment (dnjohnstone) [inline hyperspy/learn/rpca.py]: Again with the many "", there are a few examples in this file, and I won't mark them all....  In general you've made quite many 2-line things 1 line in this PR, with no change to content, was that intentional and is everything still under the max. line length?

File: hyperspy/learn/rpca.py
Comment (dnjohnstone) [inline hyperspy/learn/rpca.py]: Were these changes from ' to " also intentional? I'd always preferred ' so curious as to why if so?

File: hyperspy/learn/rpca.py
Comment (dnjohnstone) [reply hyperspy/learn/rpca.py]: In fact there are quite a few changes e.g. ' to " and regarding adding or removing brackets that look a bit like some editor you've used has made some auto changes? Do we want all of them?

**Review by dnjohnstone — CHANGES_REQUESTED** [review]
Thanks @tjof2 - just having another quick look and I get the impression your editor might have made some auto-style changes, some of which look a bit off - could you just double check that please?

File: hyperspy/learn/rpca.py
Comment (tjof2) [reply hyperspy/learn/rpca.py]: This is thanks to [black](https://black.readthedocs.io/en/stable/index.html)

File: hyperspy/learn/rpca.py
Comment (tjof2) [reply hyperspy/learn/rpca.py]: resolved most of them

**Comment by tjof2** [comment]
> Thanks @tjof2 - just having another quick look and I get the impression your editor might have made some auto-style changes, some of which look a bit off - could you just double check that please?

@dnjohnstone  I've undone most of the ones I think you were referring too. It's down to using black vs autopep8.

**Comment by tjof2** [comment]
CI failures are unrelated - all are due to `KeyError: 'PhotometricInterpretation'`

**Comment by tjof2** [comment]
@francisco-dlp / @ericpre please can we get this merged soon? I'd like to address #1159 but that would be simpler with this already in.

If not, please let me know if there's anything outstanding. Tests are failing due to the known tifffile issue.

**Comment by dnjohnstone** [comment]
> @francisco-dlp / @ericpre please can we get this merged soon? I'd like to address #1159 but that would be simpler with this already in.
> 
> If not, please let me know if there's anything outstanding. Tests are failing due to the known tifffile issue.

Just to say - I'm now +1 on merging this.

**Review by dnjohnstone — APPROVED** [review]
Thanks for modifications.

File: hyperspy/learn/mva.py
Comment (ericpre) [inline hyperspy/learn/mva.py:168]: Wondering if it would be sensible to have `'RPCA'` instead of `'RPCA_GoDec'`. To be consistent with `ORPCA` and `ORNMF`, `GoDec` should be a method and `RPCA` an algorithm? At the moment, there is only one RPCA method but they could be more in the future!

This should be considered in the context of https://github.com/hyperspy/hyperspy/issues/1678#issuecomment-312683352 too... If we want to try to more the arguments of the `blind_source_separation` and the `decomposition` method, we can keep it as it is and leave it for a separated PR and for 2.0.

**Review by ericpre — COMMENTED** [review]
This looks good to me, I have left one comment.

I would leave it to @francisco-dlp to check, since he already started this review and it knows this much better than me. He should have some time fairly soon, so he should not that long before it get merged.

File: hyperspy/learn/mva.py
Comment (tjof2) [reply hyperspy/learn/mva.py:168]: Makes sense to me, but perhaps should be addressed separately to this PR?

Comment (ericpre) [reply hyperspy/learn/mva.py:168]: Agree.

Comment (tjof2) [reply hyperspy/learn/mva.py:168]: Great - I'll keep it in mind for the near future

**Comment by francisco-dlp** [comment]
This LGTM and I don't have any further suggestions. However, codecov is not happy. This is mainly, but not only, because multiple warnings are not hit by the tests. Would it be possible to increase the coverage of the tests?

**Comment by tjof2** [comment]
@francisco-dlp  yes I'll take a look at that.

**Comment by tjof2** [comment]
@francisco-dlp I've added tests to catch the `ValueError`s and `VisibleDeprecationWarning`s that are raised in `rpca.py` and `ornmf.py`.

**Comment by tjof2** [comment]
OK @francisco-dlp I've fixed the code coverage as requested!

**Comment by francisco-dlp** [comment]
Could you update this with the latest Rnm so that the tests pass? Otherwise I'll have to go one by one to check if the only issue is the ``tifffile`` bug...

**Comment by tjof2** [comment]
@francisco-dlp it already is up-to-date, no? I merged in the following already:
```
commit dbb9821d5a9d41dd6a78af646a73c499cc0ceaa6 (upstream/RELEASE_next_minor, origin/RELEASE_next_minor, origin/HEAD, RELEASE_next_minor)
Merge: 68edb0733 4e6ec698d
Author: Francisco de la Peña <frandelapena@gmail.com>
Date:   Wed Apr 1 17:59:19 2020 +0200

    Merge pull request #2311 from dnjohnstone/copyright_update
    
    update to copyright 2020
```

**Comment by francisco-dlp** [comment]
Yes, sorry for the noise.

**Comment by tjof2** [comment]
No problem - I think it's just the tests that are queued up. It should be good I think

**Comment by francisco-dlp** [comment]
It looks that some ORNMF tests are failing. There are other tests failing (my fault), but you can fix them by merging the current Rnm

**Comment by tjof2** [comment]
Will take a look this morning!

**Comment by tjof2** [comment]
> It looks that some ORNMF tests are failing. 

@francisco-dlp  this is resolved for me locally now with the latest commit, let's wait for CI to finish, then I believe we're good to go?

> There are other tests failing (my fault), but you can fix them by merging the current Rnm

This is done.

**Comment by tjof2** [comment]
> let's wait for CI to finish

```
hyperspy\tests\mva\test_bss.py ....................................       [ 55%]
hyperspy\tests\mva\test_decomposition.py ................................ [ 56%]
hyperspy\tests\mva\test_export.py ..                                      [ 56%]
hyperspy\tests\mva\test_ornmf.py ..........                               [ 57%]
hyperspy\tests\mva\test_rpca.py ..................                        [ 57%]
```

Success! Done! :)

**Comment by tjof2** [comment]
> let's wait for CI to finish

```
hyperspy\tests\mva\test_bss.py ....................................       [ 55%]
hyperspy\tests\mva\test_decomposition.py ................................ [ 56%]
hyperspy\tests\mva\test_export.py ..                                      [ 56%]
hyperspy\tests\mva\test_ornmf.py ..........                               [ 57%]
hyperspy\tests\mva\test_rpca.py ..................                        [ 57%]
```

Success! Done! :) I tidied up the ORNMF tests to make them more consistent (i.e. always measuring against the same tolerance condition rather than hard-coded numbers that were different for each test)

**Comment by tjof2** [comment]
Spotted a bug that's been there since the start, fixed in `dac66eb` - the sparse error component in ORPCA could only be positive before now, due to the use of `np.maximum()`. I've replaced this with the correct soft-thresholding operator for l1-norm. 

Tests pass locally, so I'll wait for the CI.

```
hyperspy/tests/mva/test_bss.py ....................................             [ 55%]
hyperspy/tests/mva/test_decomposition.py ...................................... [ 56%]
hyperspy/tests/mva/test_export.py ..                                            [ 56%]
hyperspy/tests/mva/test_ornmf.py ..........                                     [ 57%]
hyperspy/tests/mva/test_rpca.py ..................                              [ 57%]
```

**Pull request merged**
